### PR TITLE
Use `bootsnap` gem to decrease dashboard boot time

### PR DESCRIPTION
**Pull request by snickell**
As investigated in https://codedotorg.atlassian.net/jira/software/c/projects/XTEAM/issues/XTEAM-396 , almost a quarter of dashboard startup time is (currently) spent on an O(N^2) filesystem search operation resolving `require` calls to paths (where N is # of gems used by dashboard).

https://github.com/Shopify/bootsnap

- Bootsnap represents a mature solution to the require path resolution problem, and was adopted as of Rails 5.1 as a default require for new projects: https://github.com/rails/rails/pull/29313
- Bootsnap is actively maintained by Shopify, and is now broadly used in the rails community
- On my system, enabling bootsnap decreased dashboard startup time by a little more than 10s.
- Additionally, Bootsnap caches .rb files precompiled as bytecode, and yaml/json serialized to binary in msgpack format. This isn't as big a benefit, but does shave a few more seconds off startup.
- Cache invalidation etc is well tested upstream
- The primary potential concern is the tmp/cache folder it creates is not automatically pruned, so long-lived environments/directories like prod/testing/staging can have a slowly growing folder size (logs problem).

As the benefit to faster startup is largest in dev-like environments, and to test the waters, this PR only installs+requires bootsnap in development and on adhoc instances.

### CODE REVIEW

**Comment by snickell** [comment]
@Hamms this might be a conversation first, on the whole Rails adopting as new project default it back in 5.1 makes me more comfortable, it does seem to be pretty mature

File: Gemfile [hunk 1]
```diff
@@ -27,6 +27,8 @@ gem 'rails-controller-testing', '~> 1.0.5'
 gem 'sprockets', github: 'wjordan/sprockets', ref: 'concurrent_asset_bundle_3.x'
 gem 'sprockets-rails', '3.3.0'
 
+gem 'bootsnap', '>= 1.14.0', group: [:development, :adhoc], require: false
```
Comment (Hamms) [inline Gemfile]: From the perspective of dashboard restart frequency, I definitely think of adhocs as being closer to our persistent long-running servers than I do our development environments. Unless there's a particular reason to enable this for adhocs, I'd prefer to minimize the number of different flavors of environment-specific exceptions we have in this file and just add this to [the `group :development` block below](https://github.com/code-dot-org/code-dot-org/pull/51636/files#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR67-R71).

**Review by Hamms — APPROVED** [review]
I'm definitely down to try this out in development, and agree that the project looks safe enough to take on as an additional dependency. It does seem likely that the potential issues with our long-running persistent servers will prevent us from adopting this across all environments for some time, but it also seems like local development is where we'll get the most benefit out of it so it's worth doing just for that.

**Comment by snickell** [comment]
> It does seem likely that the potential issues with our long-running persistent servers

Its probably not hard to fix at all for prod, we'd just have to direct the cache to a "well-known" directory like `/var/cdo-bootsnap-cache` or whatever (vs in-place like it is now, in-place is almost certainly preferred for dev), and add a cron job that does a little cleanup/pruning.

BUT, it has very little advantage for prod in the current state, IMO, so its not worth adding any extra complexity/risk.

Where it /would/ start making sense to me is if we deploy using a docker container, and as part of the docker container build we pre-boot the application and store the bootsnap-cache in-place.

1. This eliminates the long-term "cleanup" problem: every new version has a fresh perfectly matched bootsnap cache, not updating a directory in place.
2. This lowers boot time, which makes setting good auto-scaling parameters that don't under/overshoot a lot easier. Its sooooo much easier to autoscale correctly (not over or undershoot, or have weird oscillations) on a quick to start container than a slow one.

**Pull request merged**